// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Instrument protos for Forward Rate Agreement.
syntax = "proto3";

package tf_quant_finance.experimental.pricing_platform.instrument_protos;

import "tf_quant_finance/experimental/pricing_platform/instrument_protos/business_days.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/currencies.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/date.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/daycount_conventions.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/decimal.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/metadata.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/period.proto";
import "tf_quant_finance/experimental/pricing_platform/instrument_protos/rate_indices.proto";

// Floating rate term
message FloatingRateTerm {
  RateIndexType floating_rate_type = 1;
  // Reset frequency
  Period term = 2;
}

// Interest rate swap proto
message ForwardRateAgreement {
  // Whether the contract holder lends or borrows the money
  // (i.e., floating or fixed rate payer).
  // short_position = True means that the contract holder lends the money at
  // the fixed rate.
  bool short_position = 1;
  // Fixing date. Fixing date is the date at which the interest rate
  // accumulation starts.
  Date fixing_date = 2;
  // Denominated currency
  Currency currency = 3;
  // Floating rate term
  FloatingRateTerm floating_rate_term = 4;
  // Contract notional
  Decimal notional_amount = 5;
  // Fixed rate
  Decimal fixed_rate = 6;
  // Day count convention
  DayCountConvention daycount_convention = 7;
  // Business day convention
  BusinessDayConvention business_day_convention = 8;
  // Holiday calendar
  BankHolidays bank_holidays = 9;
  // Settlement days
  int32 settlement_days = 10;
  // Common metadata needed to identify the trade.
  InstrumentMetadata metadata = 11;
}
